## 1 Introduction
In this course we will be using __[Python](https://www.python.org/)__ (specifically, Python 3), one of the most popular programming language nowadays. The purpose of this document is to provide a quick tour of the syntax and of the main features of the language.

### 1.1 Obtaining Python
Python 3 is available for all major operating systems. Please [download](https://www.python.org/downloads/) the appropriate version for your computer and follow the installation instructions. It will also be useful to install the [Anaconda](https://www.anaconda.com/products/distribution) distribution platform for Python. It includes some libraries, notably __NumPy__, __SciPy__ and __matplotlib__, all of which we will use during the course, and __Jupyter__'s tools.

### 1.2 Jupyter notebooks
The present document is a __Jupyter notebook__, a document which contains both Python code and human-readable text, along with images, tables and equations. The Python code can be directly modified and interacted with by the user. To insert a new line below the current one, use `Enter` (a.k.a. `Return`) and to run the contents of the cell through the interpreter, use `Shift + Enter`. Try playing around with the cell below:

In [11]:
print("Hello word!")
pi = 3.14159
r = 1
print(f"The circumference of a circle of radius r = {r} is given by {2 * pi * r}")
pi * r**2    # '*' denotes multiplication, and '**' denotes exponentiation.

Hello word!
The circumference of a circle of radius r = 1 is given by 6.28318


3.14159

📝 Note the tags `In` for the content, or _input_, of the cell and `Out` for its result, or _output_. This output can be referred to and used in other cells through `_n`, where $ n $ is the number shown at the left. For example:

In [12]:
_1 + 2    # '+' denotes addition of numbers.

5.14159

The counter $ n $ is updated everytime the script inside the cell is run.

### 1.3 About Python
[Python](https://www.python.org/) was originally developed by Guido van Rossum starting in 1989. It is at its core an [object-oriented programming](https://en.wikipedia.org/wiki/Object-oriented_programming) (_OOP_) language. However, it also supports the main elements of the [functional programming](https://en.wikipedia.org/wiki/Functional_programming) (_FP_) and [procedural programming](https://en.wikipedia.org/wiki/Imperative_programming) paradigms.

In contrast to some other widely used languages such as C, Java or Fortran, Python is an _interpreted_ language, meaning that scripts written in Python are not compiled into machine code, but rather run by an _interpreter_.

The main advantage of interpreted languages (over compiled languages) is that they facilitate debugging and testing, since it is unnecessary to compile, link and execute after each modification of the source code. On the other hand, programs written in an interpreted language do not result in stand-alone applications, but rather require the presence of a Python interpreter in the user's computer to be run. Another disadvantage is that programs written in interpreted languages tend to have much slower execution time than their compiled equivalents, sometimes by a factor of 10 or even more.

Some other features of Python that make it suitable for teaching and learning about programming are:
* Python is _free_ and _open source_ software (_FOSS_);
* Python is available for all major operating systems and very easy to install. Programs written in any given system run without modifications in any other;
* Python's syntax is simple and very pleasant, which makes for legible code;
* Python has an extensive standard library and vibrant community of users, and its extensions are readily available.

## 2 Core Python
We shall now provide a tour of Python.

### 2.1 Variable and variable types
In programming languages, a __variable__ is used to store and manipulate a (not necessarily numerical) value. Each of these values must have a __type__, which in turn determines its properties and the operations that can be applied to it. In Python (in contrast to, say, C), variables are __typed dinamically__, meaning that not only can the _value_ of a variable change during a program, but also its _type_. To assign a value to a variable, we use the __assignment operator__ `=` in the form _variable name_ `=` _value_.

📝 Variable names may include any _alphanumeric characters_ (i.e., the characters A-Z, a-z and 0-9) together with _underscores_ \_; the initial character must be a letter. By convention we reserve uppercase initial letters to denote _classes_ and lowercase initial letters to name other objects.

📝 The type of a variable can be inspected through the function `type`.

⚠️ __Important__: In Python, simply creating a variable or assigning a new value to it does _not_ prompt the interpreter to print its value. To _print_ the value of a variable, say $ x $, use `print(x)`. To force the interpreter to _return_ the value of the variable as output, simply type its name, in this case `x`.
#### Example:

In [13]:
x = 2     # x is an integer
print(x, type(x))

x = '2'   # is now a string
print(x, type(x))

x = 2.0   # x is now a floating point number
print(x, type(x))

x

2 <class 'int'>
2 <class 'str'>
2.0 <class 'float'>


2.0

📝 The _pound_ sign __#__ is used to introduce a comment. Comments are ignored by the interpreter and can be used to  make a piece of code more readable or to clarify an intention or process.

⚠️ The following small set of keywords cannot be used to name objects in Python:

|           |            |            |           |
| :-------- | :--------- | :--------- | :------   |
|`False` 	| `def` 	 | `if`       | `raise`   |
|`None` 	| `del` 	 | `import`   | `return`  |
|`True` 	| `elif` 	 | `in` 	  | `try`     |
|`and` 	    | `else` 	 | `is` 	  | `while`   |
|`as` 	    | `except`   | `lambda`   | `with`    |
|`assert` 	| `finally`  | `nonlocal` | `yield`   |
|`break` 	| `for` 	 | `not`      |           |
|`class` 	| `from` 	 | `or` 	  |           |
|`continue` | `global`   | `pass`     |           |

📝 One can make several assignments in a single line using commas `,` as separators. This is especially useful for permuting the values of two or more variables without resorting to temporary variables:


In [ ]:
x = 1
y = 2
print(x, y)

x, y = y, x
print(x, y)

x, y = x + y, x - y
print(x, y)

## 2.1 Numerical types
Python supports three data types for storing numbers:
* __int__, or _integer_ type for integers such as $ -1 $, $ 2 $, $ 0 $ or $ 53 $;
* __float__, or _floating-point_ type for floating-point numbers (intuitively, numbers with a finite decimal expansion) such as $ 3.1415 $, $ 2.0 $ or $ -.450 $;
* __complex__, or _complex_ type for complex numbers such as $ 2 + 3j $ or $ 3.14 - 43.5 j$.

⚠️ In Python the imaginary unit is denoted by $ j $ instead of the more usual $ i $.

📝 Note that any integer may be seen as a floating-point number. In turn, any floating-point number may be seen as a complex number.

📝 The name of all of these types can also be used as _functions_ to change convert between types whenever this is possible. Similarly, we can convert all numeric types to strings using `str` (more about strings [in a moment](#strings)).

__Example:__

In [ ]:
a = 2
float_a = float(a)
complex_a = complex(a)
str_a = str(a)
print(a, type(a))
print(float_a, type(float_a))
print(complex_a, type(complex_a))
print(str_a, type(str_a))

A floating-point number to an integer, complex number or string using `int`, `complex` and `str`, respectively.

📝 Applying `int` to a floating-point number $ x $ truncates its decimal part (i.e., returns the largest integer $ <= x $).

__Example:__

In [ ]:
b = 2.0
int_b = int(b)
complex_b = complex(b)
print(b, type(b))
print(int_b, type(int_b))
print(complex_b, type(complex_b))

⚠️ A complex number can only be converted to a string. Trying to convert it to an integer or floating-point number will result in a `TypeError`.

In [ ]:
d = 2.0 + 0j
str_d = str(d)
print(str_d, type(str_d))
int_d = int(d)
print(int_d, type(int_d))

In [ ]:
float_d = float(d)
print(float_d, type(float_d))

📝 A floating-point number can also be written in the __exponential notation__ using `e` as follows:

In [ ]:
x = 1.23e2
y = 4.56e-2
z = 7.89e10
print(x)
print(y)
print(z)

The result is thus obtained by multiplying the number to the left of `e` by $ 10 $ raised to the power to the right of `e`.

### 2.2 Strings<a name="strings"></a>
A __string__ is a sequence of characters enclosed in either single `'` or double `"` quotes. Strings are __immutable__ objects, meaning that its individual characters cannot be modified during the program. Strings can be __concatenated__ using the binary operator __+__ and __sliced__ using the `[:]` notation.

⚠️ __Credits__: The content of this notebook is based on the first chapter of the textbook: Kiusalaas, J. — _Numerical Methods in Engineering with Python 3_.